src/common/com/intellij/plugins/haxe/model/HaxePackageModel.java

```diff
@@ -0,0 +1,196 @@
+    return project;
+  }
+
+  public String getName() {
+    return name;
+  }
+
+  public HaxeModel resolve(String fullyQualifiedName) {
+    return resolve(new FullyQualifiedInfo(fullyQualifiedName));
+  }
+
+  public HaxeModel resolve(FullyQualifiedInfo info) {
+    if (info.packagePath.equals(this.path)) {
+      if (info.fileName == null && info.className == null) return this;
+      HaxeFileModel file = getFileModel(info.fileName);
+      if (file != null) return file.resolve(info);
+      return null;
+    }
+    else
+      if (info.packagePath.indexOf(path) == 0 || path.isEmpty()) {
+        String searchName = path.isEmpty() ? info.packagePath : info.packagePath.substring(path.length() + 1);
+        HaxePackageModel child = getChild(searchName);
+        if (child != null) {
+          return child.resolve(info);
+        }
+        return null;
+    }
+    return null;
+  }
+
+  @Nullable
+  public HaxePackageModel getChild(@NotNull String name) {
+    if (name.isEmpty()) {
+      return this;
+    }
+    else if (name.indexOf('.') >= 0) {
+      String[] packages = StringUtils.split(name, '.');
+      HaxePackageModel result = this;
+      for (String packageName : packages) {
+        if (result == null) return null;
+        result = result.getChild(packageName);
+      }
+
+      return result;
+    }
+    else {
+      PsiDirectory directory = root.access(path.isEmpty() ? name : path + '.' + name);
+      if (directory != null) {
+        return new HaxePackageModel(project, root, name, this);
+      }
+    }
+
+    return null;
+  }
+
+  @NotNull
+  public List<HaxePackageModel> getChildren() {
+    PsiDirectory directory = root.access(path);
+    if (directory != null) {
+      return Arrays.stream(directory.getSubdirectories())
+        .map(subDirectory -> new HaxePackageModel(project, root, subDirectory.getName(), this))
+        .collect(Collectors.toList());
+    }
+    return Collections.emptyList();
+  }
+
+  @Nullable
+  public HaxeFileModel getFileModel(String fileName) {
+    PsiDirectory directory = root.access(path);
+    if (directory != null && directory.isValid()) {
+      PsiFile file = directory.findFile(fileName + ".hx");
+      if (file != null && file.isValid() && file instanceof HaxeFile) {
+        return new HaxeFileModel((HaxeFile)file);
+      }
+    }
+
+    return null;
+  }
+
+  @Nullable
+  public HaxeClassModel getClassModel(@NotNull String className) {
+    HaxeFileModel file = getFileModel(className);
+    if (file != null) {
+      return file.getClassModel(className);
+    }
+    return null;
+  }
+
+  @Override
+  public List<HaxeModel> getExposedMembers() {
+    PsiDirectory directory = root.access(path);
+    if (directory != null) {
+      PsiFile[] files = directory.getFiles();
+
+      List<HaxeModel> result = Arrays.stream(files)
+        .filter(file -> file instanceof HaxeFile)
+        .map(file -> HaxeFileModel.fromElement(file).getMainClassModel())
```

Comment:
Why only the main class?  Aren't all public classes exposed?

Reply:
No, only main classes exposed by packages, file-modules are exposing all public classes.

Reply:
I understand your point, here.  The question is whether the model is more _useful_ that way.  Is there a use case for getting only the main classes from a package, when _any_ public class can be addressed via the package FQI? (By virtue of specifying the file it exists in -- e.g. my.pkg.Class.InternalClass.)  Or, is there a need to exclude internal classes during a package search?

Reply:
Exposing mechanism mostly rely onto imports implementation that's why package exposing returning only main classes. E.g.: `import package.name.*;`

Reply:
OK.  And I'm not holding up approval based upon this.

Let's consider completion:  If I am building a completion list, I go to the index and get the related words -- indices don't care about packages.  From there, I have to re-sort the list to put the most likely candidates on top.  The natural place for me to look is in the list of imports _already_ specified, and the second place is in the package.  If only the top level classes from a package are available, then I won't have a mechanism to prioritize secondary classes easily.  Of course, I can grab the package name and use that as a filter for the index list and get around this limitation.  It's just not the _natural_ way of thinking.  

I really want the models to be *the way* to write all of the other functionality in plugin, so that we can get away from using PsiElements in our algorithms.  It will also make adding Haxe sources to the plugin easier, and I think that is the eventual goal: to be able to write (new) plugin functionality in Haxe.  To that end, I want the models to be as useful as possible.  Not that we have to write everything that we may ever need first (because that's wasteful, if we never need it), but to write the models in such a way that their usage and how to extend them is obvious.

Reply:
I understand your point, but this pull request is just first step. I don't want to jam all the things about models into this pull request. Open source is always mean ongoing work, refactoring and reinventing. I'm sure that current models implementation will be rewritten line by line about ten times till next year :-)

Reply:
>  I'm sure that current models implementation will be rewritten line by line about ten times till next year :-)

I hope not!  :)

I'd rather not have them grow "organically" though.  I'd much rather have a cohesive design at the beginning.

As I see it, Open Source work should be no less rigorous (in design, process, and coding) than "regular" work that people pay you for.  Sure, there is always refactoring and reinventing in every project and I don't expect that we can come up with a final product before we start creating it.  On the other hand, we can't ship seat-of-the-pants, "whatever works," spaghetti code either.  In the end, we're building professional products for professional developers, after all, not a cabin in the woods.